util/EIP-152/src/avx2.rs

```diff
@@ -0,0 +1,417 @@
+// Adapted from https://github.com/rust-lang-nursery/stdsimd/pull/479.
+macro_rules! _MM_SHUFFLE {
+    ($z:expr, $y:expr, $x:expr, $w:expr) => {
+        ($z << 6) | ($y << 4) | ($x << 2) | $w
+    };
+}
+
+#[inline(always)]
+unsafe fn rotate_right_32(x: __m256i) -> __m256i {
+	_mm256_shuffle_epi32(x, _MM_SHUFFLE!(2, 3, 0, 1))
+}
+
+#[inline(always)]
+unsafe fn rotate_right_24(x: __m256i) -> __m256i {
+	let rotate24 = _mm256_setr_epi8(
+		3, 4, 5, 6, 7, 0, 1, 2, 11, 12, 13, 14, 15, 8, 9, 10, 3, 4, 5, 6, 7, 0, 1, 2, 11, 12, 13,
+		14, 15, 8, 9, 10,
+	);
+	_mm256_shuffle_epi8(x, rotate24)
+}
+
+#[inline(always)]
+unsafe fn rotate_right_16(x: __m256i) -> __m256i {
+	let rotate16 = _mm256_setr_epi8(
+		2, 3, 4, 5, 6, 7, 0, 1, 10, 11, 12, 13, 14, 15, 8, 9, 2, 3, 4, 5, 6, 7, 0, 1, 10, 11, 12,
+		13, 14, 15, 8, 9,
+	);
+	_mm256_shuffle_epi8(x, rotate16)
+}
+
+#[inline(always)]
+unsafe fn rotate_right_63(x: __m256i) -> __m256i {
+	_mm256_or_si256(_mm256_srli_epi64(x, 63), add(x, x))
+}
+
+#[inline(always)]
+unsafe fn g1(a: &mut __m256i, b: &mut __m256i, c: &mut __m256i, d: &mut __m256i, m: &mut __m256i) {
+	*a = add(*a, *m);
+	*a = add(*a, *b);
+	*d = xor(*d, *a);
+	*d = rotate_right_32(*d);
+	*c = add(*c, *d);
+	*b = xor(*b, *c);
+	*b = rotate_right_24(*b);
+}
+
+#[inline(always)]
+unsafe fn g2(a: &mut __m256i, b: &mut __m256i, c: &mut __m256i, d: &mut __m256i, m: &mut __m256i) {
+	*a = add(*a, *m);
+	*a = add(*a, *b);
+	*d = xor(*d, *a);
+	*d = rotate_right_16(*d);
+	*c = add(*c, *d);
+	*b = xor(*b, *c);
+	*b = rotate_right_63(*b);
+}
+
+// Note the optimization here of leaving b as the unrotated row, rather than a.
+// All the message loads below are adjusted to compensate for this. See
+// discussion at https://github.com/sneves/blake2-avx2/pull/4
+#[inline(always)]
+unsafe fn diagonalize(a: &mut __m256i, _b: &mut __m256i, c: &mut __m256i, d: &mut __m256i) {
+	*a = _mm256_permute4x64_epi64(*a, _MM_SHUFFLE!(2, 1, 0, 3));
+	*d = _mm256_permute4x64_epi64(*d, _MM_SHUFFLE!(1, 0, 3, 2));
+	*c = _mm256_permute4x64_epi64(*c, _MM_SHUFFLE!(0, 3, 2, 1));
+}
+
+#[inline(always)]
+unsafe fn undiagonalize(a: &mut __m256i, _b: &mut __m256i, c: &mut __m256i, d: &mut __m256i) {
+	*a = _mm256_permute4x64_epi64(*a, _MM_SHUFFLE!(0, 3, 2, 1));
+	*d = _mm256_permute4x64_epi64(*d, _MM_SHUFFLE!(1, 0, 3, 2));
+	*c = _mm256_permute4x64_epi64(*c, _MM_SHUFFLE!(2, 1, 0, 3));
+}
+
+
+/// The Blake2b compression function F. See https://tools.ietf.org/html/rfc7693#section-3.2
+/// Takes as an argument the state vector `state`, message block vector `message`, offset counter, final
+/// block indicator flag `f`, and number of rounds `rounds`. The state vector provided as the first
+/// parameter is modified by the function.
+#[target_feature(enable = "avx2")]
+pub unsafe fn compress(state: &mut [u64; 8], message: [u64; 16], count: [u64; 2], f: bool, rounds: usize) {
+	// get a mutable reference to state[0..4], state[4..]
+	let (state_low, state_high) = mut_array_refs!(state, 4, 4);
```

Comment:
what is the rationale of using `array_refs/mut_array_refs` instead of `slice::split_at/slice::split_at_mut`?

Reply:
`split_at_mut` returns `&[T]` `array_refs` returns `[T; len]`

Reply:
In order to load the state into an `_m256i` I need `[u64; 4]`

Reply:
Ok, I see